src/LibUsbSharp/Internal/RundownGuard.cs

```diff
@@ -0,0 +1,441 @@
+    /// <example>
+    /// <code language="csharp"><![CDATA[
+    /// using var token = guard.AcquireSharedToken(TimeSpan.FromSeconds(1));
+    /// if (token is null)
+    ///     return; // Shutting down
+    ///
+    /// // Do work under shared protection
+    /// ]]></code>
+    /// </example>
+    public IDisposable? AcquireSharedToken(TimeSpan? timeout = null)
+    {
+        return !AcquireShared(timeout) ? null : (IDisposable)new ProtectionToken(this);
+    }
+
+    /// <summary>
+    /// Attempts to acquire shared access. Returns false if shutdown has started.
+    /// </summary>
+    /// <param name="timeout">Optional timeout. If expired, a <see cref="TimeoutException"/> is thrown.</param>
+    /// <returns>True if acquired; false if shutting down.</returns>
+    /// <exception cref="TimeoutException">Thrown if the wait exceeds <paramref name="timeout"/>.</exception>
+    public bool AcquireShared(TimeSpan? timeout = null)
+    {
+        // Start a stopwatch once so we can compute remaining time without DateTime.
+        Stopwatch? sw = timeout is null ? null : Stopwatch.StartNew();
+
+        lock (_lock)
+        {
+            while (true)
+            {
+                // If rundown/shutdown is in progress, reject new shared acquisitions.
+                if (_isShuttingDown)
+                {
+                    return false;
+                }
+
+                // Allow shared acquisition only when:
+                // - No exclusive holder is present,
+                // - No exclusive waiter exists (to avoid starvation),
+                // - Shared count has not reached the limit.
+                if (!_exclusiveHeld && _exclusiveWaiters == 0 && _activeCount < _maxSharedCount)
+                {
+                    _activeCount++;
+                    return true;
+                }
+
+                // Otherwise, wait until a state change occurs or timeout elapses.
+                AquireLock(sw, timeout);
+            }
+        }
+    }
+
+    /// <summary>
+    /// Attempts to acquire an exclusive token. Returns null if shutdown has started.
+    /// </summary>
+    /// <param name="timeout">Optional timeout. If expired, a <see cref="TimeoutException"/> is thrown.</param>
+    /// <returns>An <see cref="IDisposable"/> token that must be disposed to release the exclusive hold; null if shutting down.</returns>
+    /// <exception cref="TimeoutException">Thrown if the wait exceeds <paramref name="timeout"/>.</exception>
+    /// <example>
+    /// <code language="csharp"><![CDATA[
+    /// using var token = guard.AcquireExclusiveToken(TimeSpan.FromSeconds(5));
+    /// if (token is null)
+    ///     return; // Shutting down
+    ///
+    /// // Perform critical operation under exclusive protection
+    /// ]]></code>
+    /// </example>
+    public IDisposable? AcquireExclusiveToken(TimeSpan? timeout = null)
+    {
+        return !AcquireExclusive(timeout) ? null : (IDisposable)new ExclusiveToken(this);
```

Comment:
NITPICK: Invert conditional
```suggestion
        return AcquireExclusive(timeout) ? (IDisposable)new ExclusiveToken(this) : null;
```